pkg/restore/service_action.go

```diff
@@ -135,8 +136,14 @@ func deleteNodePorts(service *corev1api.Service) error {
 	}
 
 	for i, port := range service.Spec.Ports {
-		if !explicitNodePorts.Has(port.Name) {
-			service.Spec.Ports[i].NodePort = 0
+		if port.Name != "" {
```

Comment:
@sseago 
Could you please explain why this check is necessary, and could you please add a test case to justify this condition?

Reply:
The intent of the preserve node ports functionality is if a named port is found in the last applied configuration that matches a named port in the Service, its `nodePort` field is retained, and if a named port in the service does not have a match in that annotation, the `nodePort` field is removed. For unnamed node ports, we need to perform the same matching on the nodePort value.

The prior version of this inserted an empty entry into explicitNodePorts if there was no name. The effect was that if *any* unnamed node ports were listed in the configuration annotation, then *all* unnamed node ports were preserved, which does not match the intended goal. The fix is to track named nodeports by name and unnamed nodeports by port number. So if this port has a name, we compare it to the node port name list from the annotation, and if this port has no name, then we compare its port number to the unnamed nodeport number list from the annotation.

Reply:
For a test of this, see the modified service_action_test below. I've added a second NodePort value that doesn't match the annotation, so we see one is preserved and one is not for the `port.Name == ""` case, and the previously existing named node port test case handles testing this for the `port.Name != ""` case.

Reply:
Without this fix, the below test case modification resulted in a test failure since it preserved both unnamed ports rather than just the one that matched.